### ISSUE TITLE
Volume and delta candles

### ISSUE THREAD

**Issue by QuantFreedom1022** [project-nv/night-vision#91]
### Describe the bug

So i am not sure if this is a bug or a feature request 

i am trying to print candles based on volume and delta ... so my x axis will have random times on it like oct 14 2023 20:04:05.659 then the next one could be oct 14 2023 20:04:07.893 .... I am wondering how do i set it up so that the x axis isn't set to some type of time interval and i can make the candles start at any point with a specific width of lets say 40 pixels or something ... right now it is making my candles super super tiny ... so i want the candle size to not be dictated by the x axis

which one of the settings would i have to change ... i am guessing it could be in either 

also it seems like the answer might be here but it doesn't seems to be complete https://nightvision.dev/guide/main-comp/layout.html#mapping-functions

also if i change the data to index based each tick gets its own spot but what is interesting is the candle is perfectly spaced out ... the time is perfect on the x axis ... so really wanting to make that happen with volume based candles ... its almost as if i need to be able to turn indexbased on and off at the right time or something

this is what it is doing right now 
![image](https://github.com/project-nv/night-vision/assets/95321908/cb09ddc1-2137-48d6-b752-3528b0d0c7f9)

https://nightvision.dev/guide/api/chart-config
https://nightvision.dev/guide/api/chart-api
https://nightvision.dev/guide/api/layout-api

### Reproduction

https://github.com/QuantFreedom1022/QuantFreedom/tree/neo/nightvision/tick_data

### Javascript Framework

no-framework (vanilla-js)

### Validations

- [X] Read the [docs](https://nightvision.dev/guide/intro/getting-started.html).
- [X] Check that there isn't [already an issue](https://github.com/project-nv/night-vision/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a NightVision issue and not a framework-specific issue.
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by C451** [comment]
Seems like you need this: https://nightvision.dev/guide/main-comp/layout.html#the-problem

**Comment by QuantFreedom1022** [comment]
> Seems like you need this: https://nightvision.dev/guide/main-comp/layout.html#the-problem

yeah i tried looking into that but there doesn't seem to be much information about it ... it seems like i need to update the candle to print a new one i need the indexBased to be true ... then when i want to update the candle with ticks i need it to be false ... but i tried all of that and it still didn't work 

can you please look at this and let me know where i should be putting the change to using the indexBased true and when it should be false ... i tried to change it at chart.data.indexBased ... and it didn't work ... and i also tried to change it at chart.hub.data.indexBased and no luck ... and i also changed it in the dataload part where i load the init data and it would only print each tick as a new candle

```js
import "./style.css";
import { NightVision } from "night-vision";
import { DataLoader } from "./lib/dataLoader.js";
// import sampler from "./lib/ohlcvSampler.js";
import ticks from "./tick.json";

document.querySelector("#app").innerHTML = `
<style>
body {
    background-color: #0c0d0e;
}
</style>
<div id="chart-container"></div>
`;

let chart = new NightVision("chart-container", {
  autoResize: true,
  colors: { back: "#111113", grid: "#2e2f3055" },
  config: {},
});

let dl = new DataLoader();
dl.load((data) => {
  chart.data = data; // Set the initial data
  chart.fullReset(); // Reset tre time-range
  chart.se.uploadAndExec(); // Upload & exec scripts
});

let counter = 0;
// Setup a trade data stream
function updateCandles() {
  if (!chart.hub.mainOv) return;
  let chart_data = chart.hub.mainOv.data;
  let layout = chart.layout.main;
  if (sample(chart, chart_data, counter, layout)) {
    chart.update("data"); // New candle
    // chart.scroll(); // Scroll forward
  } else {
    chart.hub.indexBased = true;
    chart.update("data");
    chart.scroll(); // Scroll forward
    chart.hub.indexBased = false;
  }

  counter++;
}

function sample(chart, chart_data, counter, layout, vol_per_candle = 1000000) {
  let last_candle = chart_data[chart_data.length - 1];
  let current_tick = ticks[counter];
  let candle_volume = last_candle[5];
  if (!last_candle) return;
  let tick_timestamp = current_tick[0];
  let tick_price = current_tick[1];
  let tick_volume = current_tick[2];
  let new_candle_volume = tick_volume + candle_volume;

  let timestamp =
    new_candle_volume < vol_per_candle ? tick_timestamp : last_candle[0];

  if (new_candle_volume < vol_per_candle) {
    // And new zero-height candle
    let add_tick_to_candle = [
      timestamp,
      tick_price,
      tick_price,
      tick_price,
      tick_price,
      new_candle_volume,
    ];
    //callback('candle-close', symbol)
    chart_data.push(add_tick_to_candle);
    return true; // Make update('range')
  } else {
    last_candle[0] = timestamp;
    last_candle[2] = Math.max(tick_price, last_candle[2]);
    last_candle[3] = Math.min(tick_price, last_candle[3]);
    last_candle[4] = tick_price;
    last_candle[5] = 0;
    return false; // Make regular('update')
  }
}
setInterval(updateCandles, 100);
// Refernce for experiments
window.chart = chart;
```

**Comment by QuantFreedom1022** [comment]
@C451 you can check out the code in codespace ... if you could help me figure out where i would need to put indexbased to true that would be great because i wasn't able to get it to work no matter where i put it to be true or put it to be false because it seems like i need both
https://codesandbox.io/s/jovial-tess-pyf4nc?file=/.codesandbox/workspace.json

**Comment by C451** [comment]
Look into `data-aapl.json`, that will be sufficient.

**Comment by QuantFreedom1022** [comment]
> Look into `data-aapl.json` in the repo, that will be sufficient.

ok but i am not exactly sure what you would like me to do with that dataset ... it looks like every other dataset with just time and candle info ... no settings that are different ... 

https://github.com/search?q=repo%3Aproject-nv%2Fnight-vision%20data-aapl.json&type=code

i searched for it as well and only two files came up and i tested the one that did log scale and it didn't do anything 
![image](https://github.com/project-nv/night-vision/assets/95321908/b2b2e13e-5e3c-44e3-9389-e3b4fee000ff)
```js
import "./style.css";
import { NightVision } from "night-vision";
import data from "./data/data.json"

document.querySelector("#app").innerHTML = `
<style>
body {
    background-color: #0c0d0e;
}
</style>
<div id="chart-container"></div>
`;

let chart = new NightVision("chart-container", {
  autoResize: true,
  colors: { back: "#111113", grid: "#2e2f3055" },
  // scripts: [Entries, StopLosses, TakeProfits, TPFilled, SLFilled],
});
chart.data = data;
chart.data.panes[0].settings = {
  scales: {
    A: {
      log: true,
    },
  },
};
chart.se.uploadAndExec();
window.chart = chart;
```

and the other file was big tests folder 

so do you have any coding ideas i can use with this dataset ... not really sure what to do with it

**Comment by QuantFreedom1022** [comment]
After giving it some thought I realized that neither time based or index based would work ... there needs to be a third option... not sure what to call it but basically it would have to let the user build using tick data and then whenever they say they have built up enough data the close the candle and the move on to building the next ... so it would have to be a mixture of index and time ... time while building tick and index when closing the candle

I have already tried setting the index based to true and then closing the candle and then setting it back to false but that doesn't work.

So do you have any ideas on how I could help make this an option for you guys or can you help me make it because without it there's no way to do any volume based candles or tick base candles where you are building the candle off of the total amount of trades or the total amount of volume or the total amount of volume Delta

**Comment by QuantFreedom1022** [comment]
After giving it some thought I realized that neither time based or index based would work ... there needs to be other options

One would be trade-based where say for example the user is streaming binance data and they want to print a candle after every 30 trades or 50 trades or 100 trades

Then there's volume based which the user says I want to print a candle every 5 million volume or 10 million volume or whatever volume they set

And another one would be volume Delta where again if you're streaming the data you can say whenever we reach an imbalance of 500,000 by volume over the cell volume then you print a candle

  basically it would have to let the user build using tick data and then whenever they say they have built up enough data the close the candle and the move on to building the next ... exactly how time does it accept it's not based off time

I have already tried setting the index based to true and then closing the candle and then setting it back to false but that doesn't work.

So do you have any ideas on how I could help make this an option for you guys or can you help me make it because without it there's no way to do any volume based candles or tick base candles where you are building the candle off of the total amount of trades or the total amount of volume or the total amount of volume Delta

**Comment by mariusvigariu** [comment]
can't say for sure, i haven't touched that part yet .. your reproduction link says 404

**Comment by QuantFreedom1022** [comment]
@C451  ... hey just wanted to check in on this ... really could use the help in being pointed to what i need to change in order to make this work

**Comment by C451** [comment]
Hey, this project & tvjs were always for people who like to dig into the code and hack. If you want to make something tailored to your specific use-case, you'll need to work for it.

**Comment by QuantFreedom1022** [comment]
> Hey, this project & tvjs were always for people who like to dig into the code and hack. If you want to make something tailored to your specific use-case, you'll need to work for it.

ok ... but i am just needing to know if this is some type of thing that has to be added at its core or something i would have to completely change ... but its all good ... i figured it out ... and keeping in the spirit of not really wanting to help ... ill just let the next users figure it out if that's the way you want to do things!

also if its going to be that way you should probably have a massive disclaimer at the start of your documentation that your not really here to help and that if you want something done you have to figure it out yourself .... would have been good to know what i was getting into

**Comment by C451** [comment]
Well, the disclaimer is on the main page: "Time-series / candlestick charts for **PROS**." :D I wish you good luck man, I believe in your ability to figure it out.
